skxray/cdi.py

```diff
@@ -96,32 +99,321 @@ def gauss(dims, sigma):
+
+    Returns
+    -------
+    sup : array
+        support with a disk area
+    """
+    sup = np.zeros(shape_v)
+    dummy = _dist(shape_v)
+    sup[dummy < sup_radius] = 1
+    return sup
+
+
+def cdi_recon(diffracted_pattern, sample_obj, sup,
+              beta=1.15, start_avg=0.8, pi_modulus_flag='Complex',
+              sw_flag=True, sw_sigma=0.5, sw_threshold=0.1, sw_start=0.2,
+              sw_end=0.8, sw_step=10, n_iterations=1000):
+    """
+    Run reconstruction with difference map algorithm.
+
+    Parameters
+    ---------
+    diffracted_pattern : array
+        diffraction pattern from experiments
+    sample_obj : array
+        initial sample with phase
+    sup : array
+        initial support
+    beta : float, optional
+        feedback parameter for difference map algorithm.
+        default is 1.15.
+    start_avg : float, optional
+        define the point to start doing average.
+        default is 0.8.
+    pi_modulus_flag : str, optional
+        'Complex' or 'Real', defining the way to perform pi_modulus calculation.
+        default is 'Complex'.
+    sw_flag : Bool, optional
+        flag to use shrinkwrap algorithm or not.
+        default is True.
+    sw_sigma : float, optional
+        gaussian width used in sw algorithm.
+        default is 0.5.
+    sw_threshold : float, optional
+        shreshold cut in sw algorithm.
+        default is 0.1.
+    sw_start : float, optional
+        at which point to start to do shrinkwrap.
+        defualt is 0.2
+    sw_end : float, optional
+        at which point to stop shrinkwrap.
+        defualt is 0.8
+    sw_step : float, optional
+        the frequency to perform sw algorithm.
+        defualt is 10
+    n_iterations : int, optional
+        number of iterations to run.
+        default is 1000.
+
+    Returns
+    -------
+    obj_avg : array
+        reconstructed sample object
+    error_dict : dict
+        Error information for all iterations. The dict keys include
+        obj_error, diff_error and sup_error. Obj_error is a list of
+        the relative error of sample object. Diff_error is calculated as
+        the difference between new diffraction pattern and the original
+        diffraction pattern. And sup_error stores the size of the
+        sample support.
+    """
+
+    diffracted_pattern = np.array(diffracted_pattern)     # diffraction data
+
+    gamma_1 = -1/beta
+    gamma_2 = 1/beta
+
+    # get support index
+    sup_out_index = sup < 1
+
+    error_dict = {}
+    obj_error = np.zeros(n_iterations)
+    diff_error = np.zeros(n_iterations)
+    sup_error = np.zeros(n_iterations)
+
+    sup_old = np.zeros_like(diffracted_pattern)
+    obj_avg = np.zeros_like(diffracted_pattern).astype(complex)
+    avg_i = 0
+
+    time_start = time.time()
+    for n in range(n_iterations):
+        obj_old = np.array(sample_obj)
+
+        obj_a = pi_modulus(sample_obj, diffracted_pattern)
+        if pi_modulus_flag.lower() == 'real':
```

Comment:
probably worth doing this once and stashing the result above.  String operations can be expensive.